okta/okta.go

```diff
@@ -260,182 +283,136 @@ func (m *userManager) GetLoginCallbackHandler() http.HandlerFunc {
 		loginToken, err := m.cache.Put(user)
 		if err != nil {
-			err = errors.Wrap(err, "failed to cache user")
+			err = errors.Wrapf(err, "failed to cache user %s", user.Username())
+			grip.Error(err)
 			gimlet.WriteResponse(w, gimlet.MakeTextErrorResponder(err))
-			grip.Error(message.WrapError(err, message.Fields{
-				"message": "user could not be persisted in cache",
-				"op":      "GetLoginCallbackHandler",
-				"auth":    "Okta",
-			}))
 			return
 		}
 
 		m.setLoginCookie(w, loginToken)
 
-		// TODO (kim): save URI as cookie to redirect to actual requested page.
-		http.Redirect(w, r, "/", http.StatusFound)
+		http.Redirect(w, r, requestURI, http.StatusFound)
+	}
+}
+
+// getUserTokens redeems the authorization code for tokens and validates the
+// tokens.
+func (m *userManager) getUserTokens(code, nonce string) (*tokenResponse, *jwtverifier.Jwt, error) {
+	tokens, err := m.exchangeCodeForTokens(context.Background(), code)
+	if err != nil {
+		err = errors.Wrap(err, "could not redeem authorization code for tokens")
+		grip.Error(message.WrapError(err, message.Fields{
+			"endpoint": "token",
+		}))
+		return nil, nil, err
+	}
+	idToken, err := m.validateIDToken(tokens.IDToken, nonce)
+	if err != nil {
+		err = errors.Wrap(err, "invalid ID token from Okta")
+		grip.Error(err)
+		return nil, nil, err
+	}
+	if err := m.validateAccessToken(tokens.AccessToken); err != nil {
+		err = errors.Wrap(err, "invalid access token from Okta")
+		grip.Error(message.WrapError(err, message.Fields{
+			"endpoint": "introspect",
+		}))
+		return nil, nil, err
+	}
+	return tokens, idToken, nil
+}
+
+// generateUserFromInfo creates a user based on information from the userinfo
+// endpoint.
+func (m *userManager) generateUserFromInfo(tokens *tokenResponse) (gimlet.User, error) {
+	userInfo, err := m.getUserInfo(context.Background(), tokens.AccessToken)
+	if err != nil {
+		err = errors.Wrap(err, "could not retrieve user info from Okta")
+		grip.Error(message.WrapError(err, message.Fields{
+			"message":  "could not authorize user due to failure to get user info",
+			"endpoint": "userinfo",
+		}))
+		return nil, err
+	}
+	if err := m.validateGroup(userInfo.Groups); err != nil {
+		err = errors.Wrap(err, "user is not in a valid group for the organization")
+		grip.Error(message.WrapError(err, message.Fields{
+			"expected_group": m.userGroup,
+			"actual_groups":  userInfo.Groups,
+		}))
+		return nil, err
+	}
+	return makeUserFromInfo(userInfo, tokens.AccessToken, tokens.RefreshToken, m.reconciliateID), nil
+}
+
+// generateUserFromIDToken creates a user based on claims in their ID token.
+func (m *userManager) generateUserFromIDToken(tokens *tokenResponse, idToken *jwtverifier.Jwt) (gimlet.User, error) {
+	user, err := makeUserFromIDToken(idToken, tokens.AccessToken, tokens.RefreshToken, m.reconciliateID)
+	if err != nil {
+		err = errors.Wrap(err, "could not generate user from user info received from Okta")
+		grip.Error(err)
```

Comment:
It's probably unnecessary to both log and return the error in these places since the caller will probably log them anyway

Reply:
Removed.